File: quant1x/contrib/data/tdx/xdxr.go
package tdx

import (
	"fmt"

	"github.com/quant1x/quant1x/quant1x/config"
	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/level1/std"
	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/tdxproto"
	"github.com/quant1x/quant1x/quant1x/data"
	"github.com/quant1x/quant1x/quant1x/data/schema"
	"github.com/quant1x/quant1x/quant1x/encoding"
)

// UpdateXdxr 通过真实的 Level1 客户端从服务器获取除权除息数据,
// 并通过 config 包确定的文件名保存到本地缓存, 然后返回加载的切片.
// 调用者无需注册 resolver；该函数会使用 config.GetXdxrFilename 生成路径.
// 该操作会发起真实的网络请求.
func tdxFetchXdxrList(instrument data.InstrumentInfo) ([]schema.XdxrInfo, error) {
	if instrument.Type == data.SecurityTypeUnknown {
		return nil, fmt.Errorf("unknown security type for code %s", instrument.Symbol())
	}
	code := instrument.Symbol()
	// Compute filename via config package
	if len(code) != 8 {
		return nil, fmt.Errorf("invalid security code length: %s", code)
	}
	conn, release, err := GetStdConnection()
	if err != nil {
		return nil, fmt.Errorf("level1 client acquire failed: %w", err)
	}
	if release != nil {
		defer release()
	}
	if conn == nil || conn.Conn() == nil {
		return nil, fmt.Errorf("nil connection from level1 client")
	}

	msg := std.NewXdxrInfoContext(instrument)
	if err := tdxproto.TransactMessageSync(conn, msg); err != nil {
		return nil, fmt.Errorf("xdxr request failed: %w", err)
	}

	out := make([]schema.XdxrInfo, 0, len(msg.List))
	for _, it := range msg.List {
		// level1.XdxrInfo -> datasets.XdxrInfo mapping
		xi := schema.XdxrInfo{
			Date:          it.Date,
			Category:      int(it.Category),
			Name:          it.Name,
			FenHong:       it.FenHong,
			PeiGuJia:      it.PeiGuJia,
			SongZhuanGu:   it.SongZhuanGu,
			PeiGu:         it.PeiGu,
			SuoGu:         it.SuoGu,
			QianLiuTong:   it.QianLiuTong,
			HouLiuTong:    it.HouLiuTong,
			QianZongGuBen: it.QianZongGuBen,
			HouZongGuBen:  it.HouZongGuBen,
			FenShu:        it.FenShu,
			XingQuanJia:   it.XingQuanJia,
		}
		out = append(out, xi)
	}
	if len(out) == 0 {
		return nil, data.ErrNoData
	}
	fname := config.GetXdxrFilename(code)
	err = encoding.SlicesToCsv(fname, out)
	return out, err
}

func tdxGetXdxrList(sc data.InstrumentInfo) ([]schema.XdxrInfo, error) {
	if sc.Type == data.SecurityTypeUnknown {
		return nil, fmt.Errorf("unknown security type for code %s", sc.Symbol())
	}
	// 1. 确定缓存文件并读取本地缓存
	cacheFilename := config.GetXdxrFilename(sc.Symbol())
	var xdxr_list []schema.XdxrInfo
	err := encoding.CsvToSlices(cacheFilename, &xdxr_list)
	if err == nil && len(xdxr_list) > 0 {
		return xdxr_list, nil
	}

	// 2. 尝试通过网络获取数据并更新缓存
	xdxr_list, err = tdxFetchXdxrList(sc)
	if err != nil {
		return nil, err
	}

	if len(xdxr_list) > 0 {
		return xdxr_list, nil
	}

	// 3. 缓存不存在则返回无数据错误
	return nil, data.ErrNoData
}

// LoadXdxr 加载指定证券代码(如 "sh600000")的除权除息数据.
// 优先读取本地缓存, 缓存缺失时通过网络拉取并保存.
func LoadXdxr(code string) ([]schema.XdxrInfo, error) {
	return tdxGetXdxrList(data.DetectSymbol(code))
}

// DataXdxr 实现了 data.DataAdapter, 用于 XDXR 数据, 并在包初始化时注册到缓存插件中心.
type DataXdxr struct{}

func (d *DataXdxr) Kind() data.Kind { return data.BaseXdxr }
func (d *DataXdxr) Owner() string   { return data.DefaultDataProvider }
func (d *DataXdxr) Key() string     { return "xdxr" }
func (d *DataXdxr) Name() string    { return "除权除息" }
func (d *DataXdxr) Usage() string   { return "" }

func (d *DataXdxr) Print(instrument data.InstrumentInfo, dates ...data.Timestamp) {
	// No-op for now; could be extended to pretty-print loaded XDXR rows.
}

func (d *DataXdxr) Update(instrument data.InstrumentInfo, date data.Timestamp) {
	// Delegate to UpdateXdxr which fetches and writes CSV data.
	_, _ = tdxFetchXdxrList(instrument)
}

func init() {
	// Best-effort registration; ignore error if already registered.
	_ = data.Register(&DataXdxr{})
}

File: quant1x/contrib/data/tdx/xdxr.rs
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// xdxr — 除权除息数据缓存读写，与 Python contrib/data/tdx/xdxr.py 对齐
// 作为 DataAdapter 的 Rust 实现，负责网络拉取、本地 CSV 缓存和加载。

use std::sync::Arc;

use crate::data::adapter::{DataAdapter, Schema};
use crate::data::meta::instrument::Instrument;
use crate::data::meta::Timestamp;
use crate::data::schema::XdxrInfo;
use crate::data::{BaseXdxr, DEFAULT_DATA_PROVIDER};

// ============================================================
// 路径辅助
// ============================================================

/// 与 Python `_get_xdxr_filename(inst)` / C++ `xdxr_cache_filename(inst)` 对齐:
///   {cache}/xdxr/{cache_dir}/{symbol}.csv
fn xdxr_cache_filename(inst: &Instrument) -> String {
    format!(
        "{}/xdxr/{}/{}.csv",
        crate::config::default_cache_path(),
        inst.cache_dir(),
        inst.symbol()
    )
}

// ============================================================
// CSV 常量（与 Python save_xdxr 对齐）
// ============================================================

const XDXR_CSV_HEADER: &[&str] = &[
    "date",
    "category",
    "name",
    "fen_hong",
    "dividend_currency",
    "pei_gu_jia",
    "rights_currency",
    "song_zhuan_gu",
    "pei_gu",
    "suo_gu",
    "qian_liu_tong",
    "hou_liu_tong",
    "qian_zong_gu_ben",
    "hou_zong_gu_ben",
    "fen_shu",
    "xing_quan_jia",
];

fn xdxr_info_to_record(info: &XdxrInfo) -> Vec<String> {
    vec![
        info.date.clone(),
        info.category.to_string(),
        info.name.clone(),
        info.fen_hong.to_string(),
        info.dividend_currency.clone(),
        info.pei_gu_jia.to_string(),
        info.rights_currency.clone(),
        info.song_zhuan_gu.to_string(),
        info.pei_gu.to_string(),
        info.suo_gu.to_string(),
        info.qian_liu_tong.to_string(),
        info.hou_liu_tong.to_string(),
        info.qian_zong_gu_ben.to_string(),
        info.hou_zong_gu_ben.to_string(),
        info.fen_shu.to_string(),
        info.xing_quan_jia.to_string(),
    ]
}

fn record_to_xdxr_info(rec: &csv::StringRecord) -> XdxrInfo {
    XdxrInfo {
        date: rec.get(0).unwrap_or("").to_string(),
        category: rec.get(1).and_then(|s| s.parse().ok()).unwrap_or(0),
        name: rec.get(2).unwrap_or("").to_string(),
        fen_hong: rec.get(3).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        dividend_currency: rec.get(4).unwrap_or("").to_string(),
        pei_gu_jia: rec.get(5).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        rights_currency: rec.get(6).unwrap_or("").to_string(),
        song_zhuan_gu: rec.get(7).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        pei_gu: rec.get(8).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        suo_gu: rec.get(9).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        qian_liu_tong: rec.get(10).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        hou_liu_tong: rec.get(11).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        qian_zong_gu_ben: rec.get(12).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        hou_zong_gu_ben: rec.get(13).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        fen_shu: rec.get(14).and_then(|s| s.parse().ok()).unwrap_or(0.0),
        xing_quan_jia: rec.get(15).and_then(|s| s.parse().ok()).unwrap_or(0.0),
    }
}

// ============================================================
// load / save — 使用 csv crate 读写，与 Python 对齐
// ============================================================

/// 从 CSV 缓存文件加载除权除息数据
/// 与 Python `load_xdxr(inst)` 对齐
pub fn load_xdxr(inst: &Instrument) -> Vec<XdxrInfo> {
    let filename = xdxr_cache_filename(inst);
    log::debug!("Loading Xdxr data from {}", filename);

    let mut result = Vec::new();
    match std::fs::File::open(&filename) {
        Ok(f) => {
            let mut rdr = csv::ReaderBuilder::new().has_headers(true).from_reader(f);
            for row_result in rdr.records() {
                if let Ok(rec) = row_result {
                    result.push(record_to_xdxr_info(&rec));
                }
            }
        }
        Err(e) => {
            log::debug!("[xdxr] cannot open {}: {}", filename, e);
        }
    }
    result
}

/// 保存除权除息数据到 CSV 文件
/// 与 Python `save_xdxr(inst, values)` 对齐
pub fn save_xdxr(inst: &Instrument, values: &[XdxrInfo]) {
    let filename = xdxr_cache_filename(inst);
    if let Some(parent) = std::path::Path::new(&filename).parent() {
        if let Err(e) = std::fs::create_dir_all(parent) {
            log::error!("[xdxr] create_dir_all failed for {:?}: {}", parent, e);
            return;
        }
    }
    match std::fs::File::create(&filename) {
        Ok(f) => {
            let mut w = csv::Writer::from_writer(f);
            if let Err(e) = w.write_record(XDXR_CSV_HEADER) {
                log::error!("[xdxr] write header failed: {}", e);
                return;
            }
            for info in values.iter() {
                if let Err(e) = w.write_record(xdxr_info_to_record(info)) {
                    log::error!("[xdxr] write row failed: {}", e);
                }
            }
            let _ = w.flush();
        }
        Err(e) => log::error!("[xdxr] create file {} failed: {}", filename, e),
    }
}

// ============================================================
// DataXdxr — 除权除息数据适配器
// 与 Python class DataXdxr(DataAdapter) 对齐
// ============================================================

/// 除权除息数据适配器
#[derive(Debug)]
pub struct DataXdxr;

impl Schema for DataXdxr {
    fn kind(&self) -> crate::data::Kind {
        BaseXdxr
    }
    fn owner(&self) -> String {
        DEFAULT_DATA_PROVIDER.to_string()
    }
    fn key(&self) -> String {
        "xdxr".to_string()
    }
    fn name(&self) -> String {
        "除权除息".to_string()
    }
    fn usage(&self) -> String {
        "".to_string()
    }
}

impl DataAdapter for DataXdxr {
    fn print(&self, _inst: &Instrument, _dates: &[Timestamp]) {}

    fn update(&self, inst: &Instrument, _date: Timestamp) {
        let symbol = inst.symbol();
        log::debug!("[DataXdxr] update xdxr data for {}", symbol);

        // 与 C++ DataXdxr::Update / Python update_xdxr 对齐:
        //   使用批量协议拉取单只证券的除权除息数据
        let ticker = inst.market_ticker();
        let exchange = inst.exchange;

        match super::level1::std::xdxr_info::fetch_xdxr_batch(vec![(exchange, ticker.to_string())]) {
            Some(batch) => {
                let mut all_infos: Vec<XdxrInfo> = Vec::new();
                for entry in batch.list.iter() {
                    all_infos.extend(entry.list.clone());
                }
                if !all_infos.is_empty() {
                    save_xdxr(inst, &all_infos);
                } else {
                    log::debug!("[DataXdxr] no xdxr records for {}", symbol);
                }
            }
            None => {
                log::warn!("[DataXdxr] fetch_xdxr_batch returned None for {}", symbol);
            }
        }
    }
}

// ============================================================
// 注册插件 — 与 Python _data_xdxr_plugin = adapter.register(DataXdxr) 对齐
// ============================================================

pub fn init() {
    let plugin = Arc::new(DataXdxr) as Arc<dyn DataAdapter>;
    crate::data::register(plugin);
}
